sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala

```diff
@@ -46,7 +46,18 @@ class JdbcRelationProvider extends CreatableRelationProvider
     val isCaseSensitive = sqlContext.conf.caseSensitiveAnalysis
 
     val conn = JdbcUtils.createConnectionFactory(options)()
+
+    var parametersWithoutPreActions: Map[String, String] = parameters
     try {
+      options.preActions match {
+        case Some(i) =>
+          runQuery(conn, i, options)
+
+          // Remove preActions to avoid duplicate execution when writing data
+          parametersWithoutPreActions = parameters.-(JDBCOptions.JDBC_PRE_ACTIONS_STRING)
```

Comment:
The code looks weird. What's the case you're worried about here?

Reply:
As I added the comments, it is for preventing duplicate execution of `preActions` in write path.
There are two methods around here.
a) `createRelation(SQLContext, SaveMode, Map[String, String], DataFrame)`
b) `createRelation(SQLContext, Map[String, String])`

(a) is called only for writes, but (b) is called for both reads and writes
To avoid duplicate execution of preActions in writes, we need to remove preActions parameter here before passing it to (b)

Reply:
As I added the comments, it is for preventing duplicate execution of preActions in write path.
There are two methods around here.
a) `createRelation(SQLContext, SaveMode, Map[String, String], DataFrame)`
b) `createRelation(SQLContext, Map[String, String])`

(a) is called only for writes, but (b) is called for both reads and writes
To avoid duplicate execution of preActions in writes, we need to remove preActions parameter here before passing it to (b)